### PR TITLE
`clip_scalar` option to return both meshes

### PR DESCRIPTION
**Pull request by darikg**
Adds an optional argument to the `clip_scalar` filter to return both meshes. This corresponds to `GenerateClippedOutput` in VTK, a name I find a little confusing so it's just called `both` here.

### CODE REVIEW

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -285,10 +285,13 @@ def clip_scalar(dataset, scalars=None, invert=True, value=0.0, inplace=False, pr
         progress_bar : bool, optional
             Display a progress bar to indicate progress.
 
+        both : bool, optional (default False)
+            If true, also returns the complementary clipped mesh.
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:289]:
```suggestion
        both : bool, optional
            If ``True``, also returns the complementary clipped mesh.
```

File: pyvista/core/filters/data_set.py [hunk 2]
```diff
@@ -285,10 +285,13 @@ def clip_scalar(dataset, scalars=None, invert=True, value=0.0, inplace=False, pr
...
+
         Returns
         -------
-        pdata : pyvista.PolyData
-            Clipped dataset.
+        pdata : pyvista.PolyData (if `both` is False), tuple(pyvista.PolyData, pyvista.PolyData) otherwise
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:293]:
```suggestion
        :class:`pyvista.PolyData` or tuple
```

File: pyvista/core/filters/data_set.py [hunk 3]
```diff
@@ -285,10 +285,13 @@ def clip_scalar(dataset, scalars=None, invert=True, value=0.0, inplace=False, pr
...
+            Clipped dataset(s).
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:294]:
```suggestion
            Clipped dataset if ``both=False``.  If ``both=True`` then returns a tuple of both clipped datasets.
```

File: pyvista/core/filters/data_set.py [hunk 4]
```diff
@@ -300,6 +303,12 @@ def clip_scalar(dataset, scalars=None, invert=True, value=0.0, inplace=False, pr
         >>> clipped = dataset.clip_scalar(scalars="sample_point_scalars", value=100)
         >>> clipped.plot()
 
+        Get clipped meshes corresponding to the portions of the mesh above and below 100
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:306]:
```suggestion
        Get clipped meshes corresponding to the portions of the mesh above and below 100.

```

We have to have a space here to get sphinx to render it correctly.

File: pyvista/core/filters/data_set.py [hunk 5]
```diff
@@ -325,14 +334,20 @@ def clip_scalar(dataset, scalars=None, invert=True, value=0.0, inplace=False, pr
...
 
         if inplace:
-            dataset.overwrite(result)
-            return dataset
+            dataset.overwrite(result0)
+            result0 = dataset
+
+        if both:
+            result1 = _get_output(alg, oport=1)
+            return result0, result1
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:348]:
```suggestion
            return result0, _get_output(alg, oport=1)
```

**Review by akaszynski — DISMISSED** [review]
Made a few changes, but overall looks good.  Thanks for the contribution!

**Comment by akaszynski** [comment]
Please add a unit test to check `both=True`.  Looks like we're missing coverage there.

**Comment by darikg** [comment]
> Please add a unit test to check `both=True`. Looks like we're missing coverage there.

Will do!

**Comment by darikg** [comment]
Hey @akaszynski, a couple questions for you:

1. Do doctests not count towards coverage?
2. Should [this line](https://github.com/pyvista/pyvista/blob/c49bd31f0b841596cd3d7801588ff38bfc3047a4/tests/test_filters.py#L87) `assert dataset.active_scalars.min() <= clip_value` not be `max()`?
3. Do you know why `test_clip_by_scalars_filter` is marked as skip_windows and skip_mac?

**Comment by MatthewFlamm** [comment]
> 2. Should [this line](https://github.com/pyvista/pyvista/blob/c49bd31f0b841596cd3d7801588ff38bfc3047a4/tests/test_filters.py#L87) `assert dataset.active_scalars.min() <= clip_value` not be `max()`?

This test is fundamentally flawed as pointed out in https://github.com/pyvista/pyvista/pull/1475#issuecomment-872478979.  In fact, I think the line should be:

```py
assert clp.active_scalars.max() <= clip_value
```

I found that due to numerical imprecision depending on the algorithm being used, even this fails in some cases.  It is better to use something like:

```py
assert np.isclose(clp.active_scalars.max(), clip_value, rtol=5e-2)  # Note the loose tolerance
```

**Comment by adeak** [comment]
Any chance that `pytest.approx` works with inequalities?

**Comment by MatthewFlamm** [comment]
You cannot.  From [pytest docs](https://docs.pytest.org/en/6.2.x/reference.html#pytest-approx), 

>In order to avoid inconsistent behavior, TypeError is raised for >, >=, < and <= comparisons. The example below illustrates the problem:

This is the sort of test that is tricky.  The ideal state is that the clipped mesh (or the retained mesh as in the default behavior) should either have a min or max of the clip value.  In some instances, you end up with values that are slightly off of the clipped value.  I'm not sure if it will always be on one side of the clipped value.  So the inequality is not preferred IMO.  It would be better to assert that the value is close.

**Comment by adeak** [comment]
Hmm, too bad thanks.

Although should we expect floating-point errors in clipping? Clipping should either use `<` or `<=` with the clip value. Floating-point errors would imply that, say, a given value is just slightly above, so `<=` is not tripped. It should still hold true that `kept_values >= clipping_value` (or the other way around).

So I'm saying that semantically speaking, it should make sense to check for exact inequality in the test. Whether the resulting clipping is what a user would expect for the given clipping value is another matter. Does this make sense? Am I forgetting some edge cases?

**Comment by MatthewFlamm** [comment]
You are right in the general sense.  In the case of the tests, the clipped value is always in the middle of the scalar range.  I should have clarified that I was referring to the tests themselves and not in the general sense.

Thinking about this more, it might be that the imprecision I've experienced could be due to cell vs. point data.  The clip filter for cell data probably doesn't work as precisely.  In this case the inequality is the only way.

So, I'm fine keeping the tests as inequalities too.

**Comment by adeak** [comment]
>  I should have clarified that I was referring to the tests themselves and not in the general sense.

It's alright, that's how I understood it anyway :)

**Comment by akaszynski** [comment]
> 1. Do doctests not count towards coverage?

No.  I really wanted to decouple the two so we didn't rely on the documentation testing framework for unit testing, but we can consider changing this.  I agree that there's some duplication here, but at the same time unit testing and documentation example have two different objectives.

> 2. Should this line assert dataset.active_scalars.min() <= clip_value not be max()?

@MatthewFlamm gave the right answer.  Thanks

> 3. Do you know why test_clip_by_scalars_filter is marked as skip_windows and skip_mac?

Stability.  Doesn't do well on these platforms on our automated CI, and the issue is on the VTK-side.

File: tests/test_filters.py [hunk 6]
```diff
@@ -67,24 +67,42 @@ def test_clip_filter(datasets):
             else:
                 assert isinstance(clp, pyvista.UnstructuredGrid)
 
-@skip_windows
+
+# @skip_windows
```
Comment (MatthewFlamm) [inline tests/test_filters.py:71]: Is there a reason this was removed?

File: tests/test_filters.py [hunk 7]
```diff
@@ -67,24 +67,42 @@ def test_clip_filter(datasets):
...
 @skip_mac
-def test_clip_by_scalars_filter(datasets):
+@pytest.mark.parametrize('both', [False, True])
+@pytest.mark.parametrize('invert', [False, True])
+def test_clip_by_scalars_filter(both, invert):
     """This tests the clip filter on all datatypes available filters"""
+    datasets = [Sphere()]
```
Comment (MatthewFlamm) [inline tests/test_filters.py:77]: I think you should add back all the datasets to the test.  The clip filters use two different vtk algorithms depending on dataset type, so it is important that more than 1 dataset is tested here.

**Comment by darikg** [comment]
Whoops, sorry, I accidentally pushed a couple of changes I intended solely for local debugging. 

There's some failing tests now, will look into it further.

**Comment by darikg** [comment]
> > 1. Should [this line](https://github.com/pyvista/pyvista/blob/c49bd31f0b841596cd3d7801588ff38bfc3047a4/tests/test_filters.py#L87) `assert dataset.active_scalars.min() <= clip_value` not be `max()`?
> 
> This test is fundamentally flawed as pointed out in [#1475 (comment)](https://github.com/pyvista/pyvista/pull/1475#issuecomment-872478979). In fact, I think the line should be:
> 
> ```python
> assert clp.active_scalars.max() <= clip_value
> ```
> 
> I found that due to numerical imprecision depending on the algorithm being used, even this fails in some cases. It is better to use something like:
> 
> ```python
> assert np.isclose(clp.active_scalars.max(), clip_value, rtol=5e-2)  # Note the loose tolerance
> ```

Even more confusingly, the clipping behavior seems to change when setting `GenerateClippedOutput`:

```
dataset = Sphere()
dataset.point_arrays['to_clip'] = arange(dataset.n_points)
clip_val = dataset.n_points / 2

def print_range(clps):
    for clp in clps:
        c = clp.point_arrays['to_clip']
        print(f'    min={c.min()}, max={c.max()}')
    

print('Both clipped meshes, one at a time:')
print_range(
    dataset.clip_scalar('to_clip', value=clip_val, both=False, invert=invert)
    for invert in (False, True)
)
    
print('Both clipped meshes simultaneously:')
print_range(dataset.clip_scalar('to_clip', value=clip_val, both=True))
```

> Both clipped meshes, one at a time:
>     min=421, max=841
>     min=0, max=421
> Both clipped meshes simultaneously:
>     min=0, max=841
>     min=0, max=841

*Really* don't understand that.

**Comment by MatthewFlamm** [comment]
Can you test #1475 with this PR?  If that fixes it, then I think we have the extra validation to merge that PR IMO.

**Comment by darikg** [comment]
> Can you test #1475 with this PR? If that fixes it, then I think we have the extra validation to merge that PR IMO.

Doesn't change the behavior I mentioned above, but it doesn't break anything else either

**Comment by MatthewFlamm** [comment]
> Does this make sense? Am I forgetting some edge cases?

I shouldn't have tried to write in words, what can be demonstrated in code:

```py
import pyvista as pv
import numpy as np
mesh = pv.Plane()
mesh['cell_id'] = np.arange(mesh.n_points)
mesh.set_active_scalars('cell_id')
value = mesh.n_cells / 2.1
clp = mesh.clip_scalar('cell_id', value=value)
print(value)
print(clp.active_scalars.min())
print(clp.active_scalars.max())
```

yields this which does not satisfy the inequality.  This is because pyvista uses the _opposite_ default for `invert` that vtk does.

```txt
47.61904761904762
0
48
```

However, when you use the `vtk` default parameter, you get this, which is what I think @adeak is expecting, but I'm not sure.

```txt
import pyvista as pv
import numpy as np
mesh = pv.Plane()
mesh['cell_id'] = np.arange(mesh.n_points)
mesh.set_active_scalars('cell_id')
value = mesh.n_cells / 2.1
clp = mesh.clip_scalar('cell_id', value=value, invert=False)
print(value)
print(clp.active_scalars.min())
print(clp.active_scalars.max())
```

yields

```txt
47.61904761904762
48
120
```


This is why I think there should be an `np.isclose` or similar type of check.

**Comment by MatthewFlamm** [comment]
> Doesn't change the behavior I mentioned above, but it doesn't break anything else either

Bummer.  This filter is turning out to be super-unreliable.  I find the same thing with the other clip filters too unfortunately.

**Comment by darikg** [comment]
Oh, I see what the problem is. `vtkClipPolyData` with `GenerateClippedOutput True` is returning a properly clipped mesh, but all of the original points remain (just not referenced by any cells. So calling `clean` afterwards to drop unreferenced verts fixes it. Weird.

**Comment by darikg** [comment]
```python
dataset = examples.load_hexbeam()
dataset.point_arrays['to_clip'] = np.arange(dataset.n_points)
clip_value = dataset.n_points/2
clp = dataset.clip_scalar(scalars='to_clip', value=clip_value, both=False, invert=True)
print(clip_value, clp.point_arrays['to_clip'].max().item())
```
> 49.5 50

Don't think a tolerance is going to help here if we want the more stringent `assert clp.max() <= clip_value`

File: pyvista/core/filters/data_set.py [hunk 8]
```diff
@@ -318,21 +328,27 @@ def clip_scalar(dataset, scalars=None, invert=True, value=0.0, inplace=False, pr
...
-            dataset.overwrite(result)
-            return dataset
+            dataset.overwrite(result0)
+            result0 = dataset
+
+        if both:
+            result1 = _get_output(alg, oport=1)
+            if isinstance(dataset, _vtk.vtkPolyData):
+                # For some reason vtkClipPolyData with SetGenerateClippedOutput on leaves unreferenced vertices
+                result0, result1 = (r.clean() for r in (result0, result1))
```
Comment (MatthewFlamm) [inline pyvista/core/filters/data_set.py:348]: Should the result also be cleaned when `both is False`? This may make the changes in #1475 not needed?

Comment (darikg) [reply pyvista/core/filters/data_set.py:348]: AFAICT when both is False the filter doesn't return any unreferenced verts, so no need to clean (and doesn't make a difference if #1475 is merged or not)

**Comment by darikg** [comment]
Anyone know why the `clip_scalar` doctests are failing? They run locally for me.

File: Untitled.ipynb [hunk 9]
```diff
@@ -0,0 +1,250 @@
+{
```
Comment (MatthewFlamm) [inline Untitled.ipynb:1]: Remove this file

**Comment by MatthewFlamm** [comment]
> Anyone know why the clip_scalar doctests are failing? They run locally for me.

You have to add back in the `scalars=` since the dataset was reloaded.  I think the original example was flawed since the dataset was reloaded therefore there were never any active scalars.   Since you pulled in changes form the other PR, I think it is now mandatory that there are active scalars.  You'll need to remove that language from the example description as well.

**Comment by darikg** [comment]
Ok, all the tests are passing now. Two outstanding issues:

  1 - #1475 is currently merged into this PR. You may prefer to pull it out and merge it separately.
  2 - What we actually want for the test assertion. It's currently `.min() <= clip_value`. We'd prefer `.max() <= clip_value` but that runs into some numerical issues.

**Comment by akaszynski** [comment]
> 1 - #1475 is currently merged into this PR. You may prefer to pull it out and merge it separately.

That's ok.

> 2 - What we actually want for the test assertion. It's currently .min() <= clip_value. We'd prefer .max() <= clip_value but that runs into some numerical issues.

Checking into this.  At worst we can allow a wider range of acceptable values.

**Review by akaszynski — APPROVED** [review]
I've widened the acceptable values to account for VTK's inaccurate clipping.  Otherwise, LGTM.  Thanks for figuring out the vertex reference issue.
